ghcide/src/Development/IDE/Plugin/CodeAction.hs

```diff
@@ -705,6 +746,167 @@ suggestExtendImport exportsMap (L _ HsModule {hsmodImports}) Diagnostic{_range=_
+                            ImplicitPrelude{} -> True
+                            _ -> False
+                        ]
+                ]
+            | otherwise = []
+        renderUniquify HideOthers {} modName symbol =
+            "Use " <> modName <> " for " <> symbol <> ", hiding other imports"
+        renderUniquify (ToQualified qual) _ symbol =
+            "Replace with qualified: "
+                <> T.pack (moduleNameString qual)
+                <> "."
+                <> symbol
+
+occursUnqualified :: T.Text -> ImportDecl GhcPs -> Bool
+occursUnqualified symbol ImportDecl{..}
+    | isNothing ideclAs = Just False /=
+            -- I don't find this particularly comprehensible,
+            -- but HLint suggested me to do so...
+        (ideclHiding <&> \(isHiding, L _ ents) ->
+            let occurs = any ((symbol `symbolOccursIn`) . unLoc) ents
+            in isHiding && not occurs || not isHiding && occurs
+        )
+occursUnqualified _ _ = False
+
+symbolOccursIn :: T.Text -> IE GhcPs -> Bool
+symbolOccursIn symb = \case
+   IEVar _ (L _ n) -> rawIEWrapName n == T.unpack symb
+   IEThingAbs _ (L _ n) -> rawIEWrapName n == T.unpack symb
+   IEThingAll _ (L _ n) -> rawIEWrapName n == T.unpack symb
+   IEThingWith _ (L _ n) _ ents flds ->
+    rawIEWrapName n == T.unpack symb
+    || any ((== T.unpack symb) . rawIEWrapName . unLoc) ents
+    || any ((== T.unpack symb) . unpackFS . flLabel . unLoc) flds
+   IEModuleContents{} -> False
+   IEGroup{} -> False
+   IEDoc{} -> False
+   IEDocNamed{} -> False
+   XIE{} -> False
+
+targetModuleName :: ModuleTarget -> ModuleName
+targetModuleName ImplicitPrelude{} = mkModuleName "Prelude"
+targetModuleName (ExistingImp (L _ ImportDecl{..} :| _)) =
+    unLoc ideclName
+targetModuleName (ExistingImp _) =
+    error "Cannot happen!"
+
+disambiguateSymbol ::
+    DynFlags ->
+    ParsedSource ->
+    Diagnostic ->
+    T.Text ->
+    HidingMode ->
+    [Either TextEdit Rewrite]
+disambiguateSymbol df pm Diagnostic {..} (T.unpack -> symbol) = \case
+    (HideOthers hiddens0) ->
+        [ Right $ hideSymbol symbol idecl
+        | ExistingImp idecls <- hiddens0
+        , idecl <- NE.toList idecls
+        ]
+            ++ mconcat
+                [ if null imps
+                    then [Left $ hidePreludeSymbol df symbol pm]
+                    else Right . hideSymbol symbol <$> imps
+                | ImplicitPrelude imps <- hiddens0
+                ]
+    (ToQualified qualMod) ->
+        let occSym = mkVarOcc symbol
+            rdr = Qual qualMod occSym
+        in [Left $ TextEdit _range
+            $ T.pack $ showSDoc df
+            $ parenSymOcc occSym
+            $ ppr rdr]
+
+-- Couldn't make out how to add New imports by ghc-exactprint;
+-- using direct TextEdit instead.
+hidePreludeSymbol :: DynFlags -> String -> ParsedSource -> TextEdit
+hidePreludeSymbol df symbol (L _ HsModule{..}) =
+    let ran = fromJust $ srcSpanToRange $ getLoc $ last hsmodImports
```

Comment:
`ImportDecl` has an `ASTElement` instance, so you should be able to parse the new import using `parseAST` and then insert it in the list of imports. Or did you try that already?

Reply:
I once tried to insert Prelude hiding by ghc-exactprint, but get stuck.
I think we must append or prepend `import Prelude hiding  (...)` to the `hsmodImports`, but I couldn't figure out how.
I've first try to return `[LImportDecl GhcPs]` in `Rewrite`, but this won't compile because `[LImportDecl GhcPs]` was not an `ASTElement`.

Reply:
Ah, I misunderstood the purpose of `Rewrite`; it doesn't do any AST replacement, but do some replacement against the range in the sourcode with the generated AST element, right? Then I think I can find a way. Thanks!

Reply:
Yup

Reply:
Hmm. I could generate a single line of an import declaration, but couldn't make out how to prepend or append extra newlines at the level of `Rewrite`...

Reply:
... Aha! I forgot that I added `stripPrecedingNewlines` as a tentative solution. Everything works fine if one switches to `setEntryDPT`-based approach.

Reply:
OK, I rewrote `hidePreludeSymbol` using `Rewrite` somehow (much more complicated than the `TextEdit` version, though...), moved it to ExactPrint module.